File: frontend/src/api/jobs.ts
import { api } from './index'
import type { AgentRunPayload, AgentRunResponse } from './agent'
import type { ContentItem, GeneratePayload, RefinePayload } from './content'

export type JobStatus = 'queued' | 'running' | 'completed' | 'failed'
export type JobType = 'content_generation' | 'agent_run' | 'refine' | 'titles' | 'seo'

export interface JobResponse {
  id: string
  job_type: JobType
  status: JobStatus
  progress: number
  result?: Record<string, unknown> | null
  error?: string | null
  provider?: string
  model?: string
  attempts: number
  created_at?: string
  updated_at?: string
  started_at?: string
  completed_at?: string
}

export async function createContentGenerationJob(payload: GeneratePayload) {
  const { data } = await api.post<JobResponse>('/jobs/content-generation', payload)
  return data
}

export async function createAgentRunJob(payload: AgentRunPayload) {
  const { data } = await api.post<JobResponse>('/jobs/agent-run', payload)
  return data
}

export async function createRefineJob(payload: RefinePayload) {
  const { data } = await api.post<JobResponse>('/jobs/refine', payload)
  return data
}

export async function createTitlesJob(payload: Record<string, unknown>) {
  const { data } = await api.post<JobResponse>('/jobs/titles', payload)
  return data
}

export async function createSeoJob(payload: Record<string, unknown>) {
  const { data } = await api.post<JobResponse>('/jobs/seo', payload)
  return data
}

export async function getJob(jobId: string) {
  const { data } = await api.get<JobResponse>(`/jobs/${jobId}`)
  return data
}

export async function waitForJobResult<T>(
  jobId: string,
  extract: (job: JobResponse) => T | undefined,
  onUpdate?: (job: JobResponse) => void,
  timeoutMs = 360000
) {
  const started = Date.now()
  while (Date.now() - started < timeoutMs) {
    const job = await getJob(jobId)
    onUpdate?.(job)
    if (job.status === 'completed') {
      const result = extract(job)
      if (result === undefined) throw new Error('任务已完成，但结果为空')
      return result
    }
    if (job.status === 'failed') {
      throw new Error(job.error || '任务执行失败')
    }
    await new Promise(resolve => window.setTimeout(resolve, 1500))
  }
  throw new Error('任务等待超时')
}

export function extractAgentRun(job: JobResponse) {
  return job.result?.agent_run as AgentRunResponse | undefined
}

export function extractContent(job: JobResponse) {
  return job.result?.content as ContentItem | undefined
}

export function extractText(job: JobResponse) {
  return job.result?.text as string | undefined
}

File: src/api/schemas/jobs.py
from typing import Any, Literal

from pydantic import BaseModel, Field

from src.api.schemas.agent import AgentRunRequest
from src.api.schemas.content import GenerateRequest, RefineRequest, SeoRequest, TitleRequest


JobStatus = Literal["queued", "running", "completed", "failed"]
JobType = Literal["content_generation", "agent_run", "refine", "titles", "seo"]


class JobResponse(BaseModel):
    id: str
    job_type: JobType
    status: JobStatus
    progress: int = Field(0, ge=0, le=100)
    result: dict[str, Any] | None = None
    error: str | None = None
    provider: str | None = None
    model: str | None = None
    attempts: int = 0
    created_at: str | None = None
    updated_at: str | None = None
    started_at: str | None = None
    completed_at: str | None = None


class ContentGenerationJobRequest(GenerateRequest):
    pass


class AgentRunJobRequest(AgentRunRequest):
    pass


class RefineJobRequest(RefineRequest):
    pass


class TitlesJobRequest(TitleRequest):
    pass


class SeoJobRequest(SeoRequest):
    pass
